internal/controllers/topology/cluster/structuredmerge/dryrun.go

```diff
@@ -0,0 +1,301 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/cluster-api/internal/contract"
+)
+
+// getTopologyManagedFields returns metadata.managedFields entry tracking
+// server side apply operations for the topology controller.
+func getTopologyManagedFields(original client.Object) map[string]interface{} {
+	r := map[string]interface{}{}
+	if reflect.ValueOf(original).IsValid() && reflect.ValueOf(original).IsNil() {
+		return r
+	}
+
+	for _, m := range original.GetManagedFields() {
+		if m.Operation == metav1.ManagedFieldsOperationApply &&
+			m.Manager == TopologyManagerName &&
+			m.APIVersion == original.GetObjectKind().GroupVersionKind().GroupVersion().String() {
+			// NOTE: API server ensures this is a valid json.
+			err := json.Unmarshal(m.FieldsV1.Raw, &r)
+			if err != nil {
+				continue
+			}
+			break
+		}
+	}
+	return r
+}
+
+// dryRunPatch determine if the intent defined in the modified object is going to trigger
+// an actual change when running server side apply, and if this change might impact the object spec or not.
+// NOTE: This func checks if:
+// - something previously managed is missing from intent (a field has been deleted from modified)
+// - the value for a field previously managed is changing in the intent (a field has been changed in modified)
+// - the intent contains something not previously managed (a field has been added to modified).
+func dryRunPatch(ctx *dryRunInput) (hasChanges, hasSpecChanges bool) {
+	// If the func is processing a modified element of type map
+	if modifiedMap, ok := ctx.modified.(map[string]interface{}); ok {
+		// NOTE: ignoring the error in case the element wasn't in original.
+		originalMap, _ := ctx.original.(map[string]interface{})
+
+		// Process mapType/structType = granular, previously not empty.
+		// NOTE: mapType/structType = atomic is managed like scalar values down in the func;
+		// a map is atomic when the corresponding FieldV1 doesn't have info for the nested fields.
+		if len(ctx.fieldsV1) > 0 {
+			// Process all the fields the modified map
+			keys := sets.NewString()
+			hasChanges, hasSpecChanges = false, false
+			for field, fieldValue := range modifiedMap {
+				// Skip apiVersion, kind, metadata.name and metadata.namespace which are required field for a
+				// server side apply intent but not tracked into metadata.managedFields, otherwise they will be
+				// considered as a new field added to modified because not previously managed.
+				if len(ctx.path) == 0 && (field == "apiVersion" || field == "kind") {
+					continue
+				}
+				if len(ctx.path) == 1 && ctx.path[0] == "metadata" && (field == "name" || field == "namespace") {
+					continue
+				}
+
+				keys.Insert(field)
+
+				// If this isn't the root of the object and there are already changes detected, it is possible
+				// to skip processing sibling fields.
+				if len(ctx.path) > 0 && hasChanges {
+					continue
+				}
+
+				// Get the original value.
```

Comment:
nit: comment doesn't fit

Reply:
fixed

Reply:
adjusted in https://github.com/kubernetes-sigs/cluster-api/pull/6495